lib/pbench/cli/server/shell.py

```diff
@@ -63,19 +63,21 @@ def wait_for_database(db_uri: str, timeout: int):
         timeout: integer number of seconds to wait before giving up attempts to
                  connect to the database
 
-    Raises a ConnectionRefusedError on failure.
+    Raises a BadConfig exception if the DB URI specifies a host without a port,
+    and the ConnectionRefusedError enountered after the timeout.
     """
     url = urlparse(db_uri)
-    if url.scheme == "sqlite":
+    if not url.hostname:
         return
-    port = url.port if url.port is not None else 5432
+    if not url.port:
+        raise BadConfig("Database URI must contain a port number")
```

Comment:
This is "interesting" because the protocol doesn't formally require a port; the connector would normally default unless you're using an unusual port number. The problem of course is that we'd need to know the default port for each possible DB protocol in order to do the manual probe you're attempting here.

One solution would be a "formal" SQLAlchemy healthcheck, which basically amounts to something like doing the normal DB session init and trying a `select 1` query. (Which tests beyond just "something's responding on the port" to "yes, the DB is up and responding".) And it'd also work for `sqlite3` or any other lightweight local DB...

Reply:
I'd like to get this to land so we can move forward.

We are not talking about managing more than 2 configurations where the port would need to be specified: the first being the in-tree configuration used for the pbench-in-a-can container, the second being outside of the code base where it is maintained configured for the staging and production deployment (which may require a different port anyways).

This check is not trying to "know" about database semantics, just that the target database is reachable.  We could add a much deeper health check for the database, and that would probably be worth while, and could also handle the connectivity check.

But I don't see the value in waiting for that.

With the exception that I need to suss out what is happening with the functional tests.

Reply:
Fair enough: agile is good, and a check is better than none. I had approved, and I'll resolve the conversation.

I'm also going to take a look at a general solution for pulling the container log journal in a minute, having finished a round of reviews for the morning. 😆 

In the meantime, adding `podman exec pbenchinacan_3089-pbenchserver journalctl` to the `Pipeline.gy` ought to do it. (I also changed the functional test launch to add `|| true` so that failure didn't abort, but you could also put the journalctl pull in the `post.always` section before the forced remove...)